packages/smooth_app/lib/pages/folksonomy/folksonomy_manager.dart

```diff
@@ -0,0 +1,260 @@
+      } catch (e) {
+        break;
+      }
+    }
+
+    final List<FolksonomyOperation> stillPendingOperations = pendingQueue
+        .toList();
+    if (stillPendingOperations.isEmpty) {
+      await _daoTransientFolksonomy.delete(barcode);
+    } else {
+      await _daoTransientFolksonomy.put(barcode, stillPendingOperations);
+    }
+
+    if (stillPendingOperations.length == initialPendingOperations.length) {
+      // Avoid refreshing data and notifying listeners if no operations have been performed.
+      return;
+    }
+
+    try {
+      await refreshTagsFromRemote(barcode);
+    } catch (e) {
+      return;
+    }
+  }
+
+  Future<void> _syncAdd(ProductTag tag) async {
+    try {
+      final String? bearerToken = await _getBearerToken();
+      if (bearerToken == null) {
+        return;
+      }
+
+      // FIXME: The addProduct tag method does not yet have a way to add a comment.
+      final MaybeError<bool> result = await FolksonomyAPIClient.addProductTag(
+        barcode: tag.barcode,
+        key: tag.key,
+        value: tag.value,
+        bearerToken: bearerToken,
+        uriHelper: ProductQuery.uriFolksonomyHelper,
+      );
+      if (result.isError) {
+        throw Exception('${result.error}');
+      }
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  Future<void> _syncEdit(ProductTag tag) async {
+    try {
+      final String? bearerToken = await _getBearerToken();
+      if (bearerToken == null) {
+        return;
+      }
+
+      final MaybeError<bool> result =
+          await FolksonomyAPIClient.updateProductTag(
+            barcode: tag.barcode,
+            key: tag.key,
+            value: tag.value,
+            version: tag.version,
+            bearerToken: bearerToken,
+            uriHelper: ProductQuery.uriFolksonomyHelper,
+          );
+      if (result.isError) {
+        throw Exception('${result.error}');
+      }
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  Future<void> _syncDelete(ProductTag tag) async {
+    try {
+      final String? bearerToken = await _getBearerToken();
+      if (bearerToken == null) {
+        return;
+      }
+
+      final MaybeError<bool> result =
+          await FolksonomyAPIClient.deleteProductTag(
+            barcode: tag.barcode,
+            key: tag.key,
+            version: tag.version,
+            bearerToken: bearerToken,
+            uriHelper: ProductQuery.uriFolksonomyHelper,
+          );
+      if (result.isError) {
+        throw Exception('${result.error}');
+      }
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  Future<void> refreshTagsFromRemote(String barcode) async {
+    final Map<String, ProductTag> tags =
+        await FolksonomyAPIClient.getProductTags(
+          barcode: barcode,
+          uriHelper: ProductQuery.uriFolksonomyHelper,
+        );
+    await _daoFolksonomy.put(barcode, tags.values.toList());
+
+    _localDatabase.notifyListeners();
+  }
+
+  Future<String?> _getBearerToken() async {
```

Comment:
```suggestion
  Future<String> _getBearerToken() async {
```
Either you return a not null `String` or you throw an exception.